pkg/auth/auth.go

```diff
@@ -15,21 +16,96 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+type HeaderAuthName string
+
+func (h HeaderAuthName) String() string { return string(h) }
+
 // XRegistryAuthHeader is the key to the encoded registry authentication
 // configuration in an http-request header.
-const XRegistryAuthHeader = "X-Registry-Auth"
+const XRegistryAuthHeader HeaderAuthName = "X-Registry-Auth"
+
+// XRegistryConfigHeader is the key to the encoded registry authentication
+// configuration in an http-request header.
+const XRegistryConfigHeader HeaderAuthName = "X-Registry-Config"
+
+// GetCredentials queries the http.Request for X-Registry-.* headers and extracts
+// the necessary authentication information for libpod operations
+func GetCredentials(r *http.Request) (*types.DockerAuthConfig, string, HeaderAuthName, error) {
+	has := func(key HeaderAuthName) bool { hdr, found := r.Header[string(key)]; return found && len(hdr) > 0 }
+	switch {
+	case has(XRegistryConfigHeader):
+		c, f, err := getConfigCredentials(r)
+		return c, f, XRegistryConfigHeader, err
+	case has(XRegistryAuthHeader):
+		c, f, err := getAuthCredentials(r)
+		return c, f, XRegistryConfigHeader, err
```

Comment:
That should be XRegistryAuthHeader.

It looks like the 2nd return value is only used for error messages. Can we simplify further by embedding it in the error message here?

Reply:
Fixed bug.  I could move error handling, but it would really break segregation of duties as I would need to format the http body here.